### DISCUSSION TITLE
Split issue identifier

### DISCUSSION THREAD

**Discussion by mizaki** [comictagger/comictagger#544] [Ideas]
I believe one of the main problems with supporting series only talkers and to cover #234 is that `ii` does a joint series and issue search and identification. I propose splitting the `series` and `issue` parts and also remove `ComicSeries` and use `GenericMetadata` for both series and issue internally.

I know you've been against a `hasIssues` talker variable before and `ii` not needing to know but for the issue window and tagging without an issue number etc., I'm not sure how it's possible without some indication of capability from the talker.

With this knowledge (and an option to cover #234) `ii` could operate with series only talkers. It would still be a requirement that an issue talker needs an issue number.

How does that sound?

**Comment by mizaki** [comment]
An additional idea along with using `GenericMetadata` is to add an `hasIssues` to that. With the addition of the `on click` functionality, the series window's buttons can be disabled/enabled per comic series. Using `GenericMetadata` internally for all should make it much easier to use series only data.

Talkers could then mark `GenericMetadata` as having no issue either by default (the source does not support it) or even check per series.

How to go about the option(s) for the user could come down to: `Tag with series information if no issue information is available` should be global, per talker, or both. Or even always tag with whatever information there is.

**Reply by lordwelch** [reply]
So I don't want to have a has_issues attribute because if a series, whether manga, standard comics or something else, doesn't have any issues then I don't care about it, a series is a container for issues if it's empty (no issues) then there's nothing useful about it and it might as well not exist. I dare **_anyone_** to find me a _completed_ comic with _no_ issues or a _completed_ manga with _no_ chapters.

If we go on the assumption that, if a series is found, it has to have at least one issue published. Then it makes this a lot simpler: we just return 1 issue when we don't have any more information and for those we shouldn't even need to return an issue number just leave it set to `None`. The issue identifier doesn't ever check the issue number of the returned issue, it just uses the provided issue number as a search parameter.

If I don't know how many issues a series has that is very similar to not having an issue number and trying to search, so if instead we solve that it will apply to everything, instead of making a separate option that only works in a specific edge-case.

> I propose splitting the series and issue parts

The issue identifier definitely needs a rewrite, it's basically a giant single function masquerading as a class.

> remove ComicSeries and use GenericMetadata for both series and issue internally.

This just makes everything more complicated because now we have no separation, so when I ask for a list of series and one has an issue number and an issue id what do I do with it? Fully separating issues and series out makes much more sense because then when searching for a series there wouldn't be any issue information to make it more complicated. GenericMetadata is a random collection of any properties that would fit in any metadata scheme (not exhaustive, at least currently) and is a random combination of metadata that only applies to a series and metadata that only applies to an issue and the things that would apply to both wouldn't usually be the same if you were to fill it out accurately. Completely getting rid of GenericMetadata and just passing around a tuple of (Series, Issue) is much more likely to happen.

**Comment by mizaki** [comment]
> So I don't want to have a has_issues attribute because if a series, whether manga, standard comics or something else, doesn't have any issues then I don't care about it, a series is a container for issues if it's empty (no issues) then there's nothing useful about it and it might as well not exist. I dare **_anyone_** to find me a _completed_ comic with _no_ issues or a _completed_ manga with _no_ chapters.
> 
This is information source versus actual comic. The information for the series is useful (publisher, artist/writer (manga is usually one person), characters and so on). It could be argued for manga that the chapter (and to a lesser extend, volume) information is of little value. Excluding MangaUpdates, Anilist, MyAnimeList (it has manga too) does not (currently) seem ideal. (Kitsu may be the general answer but may never have the coverage of MU/MAL.)

The question is: How can MU (et al) be included with CT?
As I've mentioned before, I see
- The MU talker can return the series data as an issue (easier now using GMD for issues. I have a new version almost ready that does this.)
- How to address the GUI problem of not showing the issue window. (This is the blocker to my mind.)

> If we go on the assumption that, if a series is found, it has to have at least one issue published. Then it makes this a lot simpler: we just return 1 issue when we don't have any more information and for those we shouldn't even need to return an issue number just leave it set to `None`. The issue identifier doesn't ever check the issue number of the returned issue, it just uses the provided issue number as a search parameter.
> 
That seems like a bad idea as the extra information (publication date, synopsis, characters, etc.) can all be wrong. An option such as `Tag with series information only if issue number is not found` would be more sensible. And reuse what was parsed from the file name (issue number). This would cover #234 

> If I don't know how many issues a series has that is very similar to not having an issue number and trying to search, so if instead we solve that it will apply to everything, instead of making a separate option that only works in a specific edge-case.
> 

> > I propose splitting the series and issue parts
> 
> The issue identifier definitely needs a rewrite, it's basically a giant single function masquerading as a class.
> 
> > remove ComicSeries and use GenericMetadata for both series and issue internally.
> 
> This just makes everything more complicated because now we have no separation, so when I ask for a list of series and one has an issue number and an issue id what do I do with it? Fully separating issues and series out makes much more sense because then when searching for a series there wouldn't be any issue information to make it more complicated. GenericMetadata is a random collection of any properties that would fit in any metadata scheme (not exhaustive, at least currently) and is a random combination of metadata that only applies to a series and metadata that only applies to an issue and the things that would apply to both wouldn't usually be the same if you were to fill it out accurately. Completely getting rid of GenericMetadata and just passing around a tuple of (Series, Issue) is much more likely to happen.

The separation comes from the function. A series function only looks for series data. It could be in `ComicSeries` or `GenericMetadata` or etc., it doesn't matter. My reason for this suggestion is something like:

1. `ii` gets series info as `gmd`
2. `ii` searches for issue info
3a. If it's found a `gmd` comes back with the series and issue data
3b. If not, `ii` can write the series `gmd` (if the option is set etc.)

As to getting rid of GenericMetadata. At some point `title` needs to be mapped to the metadata container `title` so a definitive starting point is needed. A metadata plugin being able to expand a genericmetadata object for storing unsupported data from reading would make sense so it could be written back as is but that's another topic.

**Reply by lordwelch** [reply]
> That seems like a bad idea as the extra information (publication date, synopsis, characters, etc.) can all be wrong.

If it's wrong, then why is it being returned? I don't want to just arbitrarily apply all series information to a comic, that's not a great idea and it is not what is currently done. Currently we query a source by providing it inputs and then applying what it returns, the only check anywhere is the series name when searching, for 2 reasons:
1. ComicVine always returns results and we don't want to show everything on CV for every search
1. Many names of series are similar to other series and if we can trim down some of the extra results to get a single match that is better

CT assumes whatever is returned from a talker is correct, if it return multiple results (specifically for the series) we let the user decide which result to use. CT doesn't technically make any other decisions about the metadata that is returned.

>  An option such as `Tag with series information only if issue number is not found` would be more sensible. And reuse what was parsed from the file name (issue number). This would cover #234

We already have an algorithm in place to preserve metadata from any given source and it works well. I don't see why it needs to change.

1. Parse metadata from filename
1. Read metadata from known styles. Any metadata defined here overrides metadata parsed from the filename
1. Apply metadata explicitly defined (either through the GUI or on the CLI through the `-m` option). Any metadata defined here overrides metadata parsed from the filename and metadata read from the known style
1. Apply metadata from a talker. Any metadata defined here overrides metadata parsed from the filename and metadata read from the known style and explicit metadata

> ii gets series info as gmd
ii searches for issue info
3a. If it's found a gmd comes back with the series and issue data
3b. If not, ii can write the series gmd (if the option is set etc.)

How about we amend this:
1. II gets series information
1. II searches for issue info
   1. The talker returns whatever is correct for that issue. If it returns the exact same data for all issues that's fine!

Lets do an over explained example:
1. Open `Full Metal Alchemist #5 (2001).cbz`
1. CT parses the series as `Full Metal Alchemist` and the issue number as `5`
1. We have opened up a search for 'Full Metal Alchemist' on a MangaUpdates talker
1. Select the correct series
1. The MangaUpdates talker doesn't find any information on chapter/volume information (even though the site reports that it has [27 volumes])
1. The MangaUpdates talker returns a single result with whatever information it does find, mapped as it deems fit for tagging a CBZ.
1. See that the MangaUpdates talker has supplied only one issue
1. Select the issue
1. Metadata is applied to the current metadata, any metadata set by the MangaUpdates talker overrides any metadata already existing.
1. Note that the issue number remains as 5 because the MangaUpdates talker returned metadata without an issue number

We can modify this slightly for auto-tagging:
1. Open `Full Metal Alchemist #5 (2001).cbz`
1. CT parses the series as `Full Metal Alchemist` and the issue number as `5`
1. CT searches for 'Full Metal Alchemist' on a MangaUpdates talker
1. CT selects the first series as it is an exact match
1. CT Searches for issue number 5
1. The MangaUpdates talker doesn't find any information on chapter/volume information (even though the site reports that it has [27 volumes])
1. The MangaUpdates talker returns a single result with whatever information it does find, mapped as it deems fit for tagging a CBZ.
1. CT only gets 1 result so marks it as a successful search
1. Metadata is applied to the current metadata, any metadata set by the MangaUpdates talker overrides any metadata already existing.
1. Note that the issue number remains as 5 because the MangaUpdates talker returned metadata without an issue number

[27 volumes]: https://www.mangaupdates.com/series/8q91xb4/hagane-no-renkinjutsushi

**Comment by mizaki** [comment]
In terms of paths for the new `ii` to take with a talker:
1. Search for series with issue number [and year]^
2. Search for series (as now)
2.1 Search for issues with number [and year] (as now)
2.2 Check hash from API against local cover hash (if API provides a suitable hash)
2.3 If 2.2 fails (no hash from API etc.) download cover to generate hash (as now)

^ This I believe would require a new function added to `ComicTalker` along the lines of `search_for_series_and_issue`. Then if that comes back as `NotImplemeted` the traditional path can be taken.

**Reply by lordwelch** [reply]
Trying to figure out implementation is never going to work until we know what we want it to look like. First what information does [metron_talker](https://github.com/comictagger/metron_talker), [gcd_talker](https://github.com/comictagger/gcd_talker) or any other plugin want that it currently doesn't get, or does not get in an optimal way? Second what information do any of these plugins have to provide that it is not able to or isn't able to in an optimal way?
